File: app/src/main/java/com/example/bloodbank/UpdateProfileActivity.java
package com.example.bloodbank;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.FileProvider;

import android.app.DatePickerDialog;
import android.app.ProgressDialog;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.view.MenuItem;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.HashMap;
import java.util.Locale;

import de.hdodenhof.circleimageview.CircleImageView;

public class UpdateProfileActivity extends AppCompatActivity {

    private Toolbar toolbar;
    private CircleImageView profileImage;
    private MaterialButton changeImageButton, updateButton;
    private TextInputLayout nameLayout, phoneLayout, addressLayout, birthdateLayout;
    private TextInputEditText nameInput, phoneInput, addressInput, birthdateInput;

    private Uri resultUri;
    private ProgressDialog loader;
    private DatabaseReference userRef;
    private File currentPhotoFile;
    private static final int REQUEST_IMAGE_CAPTURE = 1;
    private static final int REQUEST_IMAGE_PICK = 2;
    private static final String APP_DIRECTORY = "BloodBank";
    private static final String IMAGE_DIRECTORY = "ProfileImages";

    private Calendar calendar;
    private SimpleDateFormat dateFormat;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_update_profile);

        // Initialize toolbar
        toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle("Update Profile");
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);

        // Initialize views
        initializeViews();
        
        // Initialize Firebase references
        userRef = FirebaseDatabase.getInstance().getReference()
                .child("users").child(FirebaseAuth.getInstance().getCurrentUser().getUid());

        // Load current user data
        loadCurrentData();
        
        // Set up date picker
        setupDatePicker();
        
        // Set up image picker
        setupImagePicker();
        
        // Set up update button
        setupUpdateButton();
    }

    private void initializeViews() {
        profileImage = findViewById(R.id.profileImage);
        changeImageButton = findViewById(R.id.changeImageButton);
        updateButton = findViewById(R.id.updateButton);
        
        nameLayout = findViewById(R.id.nameLayout);
        phoneLayout = findViewById(R.id.phoneLayout);
        addressLayout = findViewById(R.id.addressLayout);
        birthdateLayout = findViewById(R.id.birthdateLayout);
        
        nameInput = findViewById(R.id.nameInput);
        phoneInput = findViewById(R.id.phoneInput);
        addressInput = findViewById(R.id.addressInput);
        birthdateInput = findViewById(R.id.birthdateInput);

        loader = new ProgressDialog(this);
        calendar = Calendar.getInstance();
        dateFormat = new SimpleDateFormat("dd/MM/yyyy", Locale.getDefault());
    }

    private void loadCurrentData() {
        userRef.addListenerForSingleValueEvent(new com.google.firebase.database.ValueEventListener() {
            @Override
            public void onDataChange(@NonNull com.google.firebase.database.DataSnapshot snapshot) {
                if (snapshot.exists()) {
                    String name = snapshot.child("name").getValue(String.class);
                    String phone = snapshot.child("phonenumber").getValue(String.class);
                    String address = snapshot.child("address").getValue(String.class);
                    String birthdate = snapshot.child("birthdate").getValue(String.class);
                    String profileImagePath = snapshot.child("profileImagePath").getValue(String.class);

                    nameInput.setText(name);
                    phoneInput.setText(phone);
                    addressInput.setText(address);
                    birthdateInput.setText(birthdate);

                    if (profileImagePath != null && !profileImagePath.isEmpty()) {
                        try {
                            File imgFile = new File(profileImagePath);
                            if (imgFile.exists()) {
                                Bitmap myBitmap = BitmapFactory.decodeFile(imgFile.getAbsolutePath());
                                profileImage.setImageBitmap(myBitmap);
                            }
                        } catch (Exception e) {
                            Toast.makeText(UpdateProfileActivity.this, "Error loading profile image", 
                                    Toast.LENGTH_SHORT).show();
                        }
                    }
                }
            }

            @Override
            public void onCancelled(@NonNull com.google.firebase.database.DatabaseError error) {
                Toast.makeText(UpdateProfileActivity.this, "Error loading profile: " + error.getMessage(),
                        Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void setupDatePicker() {
        birthdateInput.setOnClickListener(v -> {
            DatePickerDialog datePickerDialog = new DatePickerDialog(
                    UpdateProfileActivity.this,
                    (view, year, month, dayOfMonth) -> {
                        calendar.set(Calendar.YEAR, year);
                        calendar.set(Calendar.MONTH, month);
                        calendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                        birthdateInput.setText(dateFormat.format(calendar.getTime()));
                    },
                    calendar.get(Calendar.YEAR),
                    calendar.get(Calendar.MONTH),
                    calendar.get(Calendar.DAY_OF_MONTH)
            );
            datePickerDialog.show();
        });
    }

    private void setupImagePicker() {
        changeImageButton.setOnClickListener(v -> {
            Intent intent = new Intent(Intent.ACTION_PICK);
            intent.setType("image/*");
            startActivityForResult(intent, REQUEST_IMAGE_PICK);
        });
    }

    private File createImageFile() throws IOException {
        // Create app directory if it doesn't exist
        File appDir = new File(getFilesDir(), APP_DIRECTORY);
        if (!appDir.exists()) {
            appDir.mkdirs();
        }

        // Create images directory if it doesn't exist
        File imageDir = new File(appDir, IMAGE_DIRECTORY);
        if (!imageDir.exists()) {
            imageDir.mkdirs();
        }

        // Create image file
        String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault()).format(System.currentTimeMillis());
        String imageFileName = "PROFILE_" + timeStamp + ".jpg";
        return new File(imageDir, imageFileName);
    }

    private void setupUpdateButton() {
        updateButton.setOnClickListener(v -> {
            String name = nameInput.getText().toString().trim();
            String phone = phoneInput.getText().toString().trim();
            String address = addressInput.getText().toString().trim();
            String birthdate = birthdateInput.getText().toString().trim();

            if (TextUtils.isEmpty(name)) {
                nameLayout.setError("Name is required");
                return;
            }
            if (TextUtils.isEmpty(phone)) {
                phoneLayout.setError("Phone number is required");
                return;
            }
            if (TextUtils.isEmpty(address)) {
                addressLayout.setError("Address is required");
                return;
            }
            if (TextUtils.isEmpty(birthdate)) {
                birthdateLayout.setError("Birthdate is required");
                return;
            }

            loader.setMessage("Updating profile...");
            loader.setCanceledOnTouchOutside(false);
            loader.show();

            HashMap<String, Object> userMap = new HashMap<>();
            userMap.put("name", name);
            userMap.put("phonenumber", phone);
            userMap.put("address", address);
            userMap.put("birthdate", birthdate);

            if (currentPhotoFile != null) {
                userMap.put("profileImagePath", currentPhotoFile.getAbsolutePath());
            }

            userRef.updateChildren(userMap).addOnCompleteListener(task -> {
                loader.dismiss();
                if (task.isSuccessful()) {
                    Toast.makeText(UpdateProfileActivity.this, "Profile updated successfully",
                            Toast.LENGTH_SHORT).show();
                    finish();
                } else {
                    Toast.makeText(UpdateProfileActivity.this, "Error updating profile: " + task.getException().getMessage(),
                            Toast.LENGTH_SHORT).show();
                }
            });
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK) {
            if (requestCode == REQUEST_IMAGE_PICK && data != null) {
                resultUri = data.getData();
                try {
                    // Create a new file to save the image
                    currentPhotoFile = createImageFile();
                    
                    // Copy the selected image to the new file
                    Bitmap bitmap = MediaStore.Images.Media.getBitmap(getContentResolver(), resultUri);
                    FileOutputStream out = new FileOutputStream(currentPhotoFile);
                    bitmap.compress(Bitmap.CompressFormat.JPEG, 90, out);
                    out.close();
                    
                    // Display the image
                    profileImage.setImageBitmap(bitmap);
                } catch (IOException e) {
                    Toast.makeText(this, "Error saving image: " + e.getMessage(), Toast.LENGTH_SHORT).show();
                }
            }
        }
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
} 

File: app/src/main/java/com/example/bloodbank/UpdateProfileActivity.kt
package com.example.bloodbank

import android.app.DatePickerDialog
import android.app.ProgressDialog
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.view.MenuItem
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.bumptech.glide.Glide
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import de.hdodenhof.circleimageview.CircleImageView
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale

/**
 * UpdateProfileActivity - Allows users to update their profile information.
 * Uses modern Android APIs like registerForActivityResult for image picking.
 */
class UpdateProfileActivity : AppCompatActivity() {
    
    // Views
    private lateinit var toolbar: Toolbar
    private lateinit var profileImage: CircleImageView
    private lateinit var changeImageButton: MaterialButton
    private lateinit var updateButton: MaterialButton
    private lateinit var nameLayout: TextInputLayout
    private lateinit var phoneLayout: TextInputLayout
    private lateinit var addressLayout: TextInputLayout
    private lateinit var birthdateLayout: TextInputLayout
    private lateinit var nameInput: TextInputEditText
    private lateinit var phoneInput: TextInputEditText
    private lateinit var addressInput: TextInputEditText
    private lateinit var birthdateInput: TextInputEditText
    
    // State
    private var currentPhotoFile: File? = null
    private lateinit var loader: ProgressDialog
    private val calendar = Calendar.getInstance()
    private val dateFormat = SimpleDateFormat("dd/MM/yyyy", Locale.getDefault())
    
    // Firebase
    private val userRef by lazy {
        FirebaseDatabase.getInstance().getReference()
            .child("users")
            .child(FirebaseAuth.getInstance().currentUser?.uid ?: "")
    }
    
    // Modern image picker using Activity Result API
    private val imagePickerLauncher = registerForActivityResult(
        ActivityResultContracts.GetContent()
    ) { uri: Uri? ->
        uri?.let { handleImageSelection(it) }
    }
    
    companion object {
        private const val APP_DIRECTORY = "BloodBank"
        private const val IMAGE_DIRECTORY = "ProfileImages"
    }
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_update_profile)
        
        // Initialize views
        initializeViews()
        
        // Setup toolbar
        setupToolbar()
        
        // Load current user data
        loadCurrentData()
        
        // Setup date picker
        setupDatePicker()
        
        // Setup image picker
        setupImagePicker()
        
        // Setup update button
        setupUpdateButton()
    }
    
    private fun initializeViews() {
        toolbar = findViewById(R.id.toolbar)
        profileImage = findViewById(R.id.profileImage)
        changeImageButton = findViewById(R.id.changeImageButton)
        updateButton = findViewById(R.id.updateButton)
        
        nameLayout = findViewById(R.id.nameLayout)
        phoneLayout = findViewById(R.id.phoneLayout)
        addressLayout = findViewById(R.id.addressLayout)
        birthdateLayout = findViewById(R.id.birthdateLayout)
        
        nameInput = findViewById(R.id.nameInput)
        phoneInput = findViewById(R.id.phoneInput)
        addressInput = findViewById(R.id.addressInput)
        birthdateInput = findViewById(R.id.birthdateInput)
        
        loader = ProgressDialog(this)
    }
    
    private fun setupToolbar() {
        setSupportActionBar(toolbar)
        supportActionBar?.apply {
            title = "Update Profile"
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowHomeEnabled(true)
        }
    }
    
    private fun loadCurrentData() {
        userRef.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (snapshot.exists()) {
                    nameInput.setText(snapshot.child("name").getValue(String::class.java))
                    phoneInput.setText(snapshot.child("phonenumber").getValue(String::class.java))
                    addressInput.setText(snapshot.child("address").getValue(String::class.java))
                    birthdateInput.setText(snapshot.child("birthdate").getValue(String::class.java))
                    
                    val profileImagePath = snapshot.child("profileImagePath").getValue(String::class.java)
                    profileImagePath?.let { loadProfileImage(it) }
                }
            }
            
            override fun onCancelled(error: DatabaseError) {
                Toast.makeText(
                    this@UpdateProfileActivity,
                    "Error loading profile: ${error.message}",
                    Toast.LENGTH_SHORT
                ).show()
            }
        })
    }
    
    private fun loadProfileImage(imagePath: String) {
        try {
            val imgFile = File(imagePath)
            if (imgFile.exists()) {
                val bitmap = BitmapFactory.decodeFile(imgFile.absolutePath)
                profileImage.setImageBitmap(bitmap)
            }
        } catch (e: Exception) {
            Toast.makeText(this, "Error loading profile image", Toast.LENGTH_SHORT).show()
        }
    }
    
    private fun setupDatePicker() {
        birthdateInput.setOnClickListener {
            DatePickerDialog(
                this,
                { _, year, month, dayOfMonth ->
                    calendar.set(Calendar.YEAR, year)
                    calendar.set(Calendar.MONTH, month)
                    calendar.set(Calendar.DAY_OF_MONTH, dayOfMonth)
                    birthdateInput.setText(dateFormat.format(calendar.time))
                },
                calendar.get(Calendar.YEAR),
                calendar.get(Calendar.MONTH),
                calendar.get(Calendar.DAY_OF_MONTH)
            ).show()
        }
    }
    
    private fun setupImagePicker() {
        changeImageButton.setOnClickListener {
            imagePickerLauncher.launch("image/*")
        }
    }
    
    private fun handleImageSelection(uri: Uri) {
        try {
            // Create a new file to save the image
            currentPhotoFile = createImageFile()
            
            // Copy the selected image to the new file
            val bitmap = MediaStore.Images.Media.getBitmap(contentResolver, uri)
            FileOutputStream(currentPhotoFile).use { out ->
                bitmap.compress(Bitmap.CompressFormat.JPEG, 90, out)
            }
            
            // Display the image
            profileImage.setImageBitmap(bitmap)
        } catch (e: IOException) {
            Toast.makeText(this, "Error saving image: ${e.message}", Toast.LENGTH_SHORT).show()
        }
    }
    
    private fun createImageFile(): File {
        // Create app directory if it doesn't exist
        val appDir = File(filesDir, APP_DIRECTORY).apply {
            if (!exists()) mkdirs()
        }
        
        // Create images directory if it doesn't exist
        val imageDir = File(appDir, IMAGE_DIRECTORY).apply {
            if (!exists()) mkdirs()
        }
        
        // Create image file
        val timeStamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault())
            .format(System.currentTimeMillis())
        val imageFileName = "PROFILE_$timeStamp.jpg"
        
        return File(imageDir, imageFileName)
    }
    
    private fun setupUpdateButton() {
        updateButton.setOnClickListener {
            if (validateInputs()) {
                updateProfile()
            }
        }
    }
    
    private fun validateInputs(): Boolean {
        val name = nameInput.text.toString().trim()
        val phone = phoneInput.text.toString().trim()
        val address = addressInput.text.toString().trim()
        val birthdate = birthdateInput.text.toString().trim()
        
        // Clear previous errors
        nameLayout.error = null
        phoneLayout.error = null
        addressLayout.error = null
        birthdateLayout.error = null
        
        return when {
            name.isEmpty() -> {
                nameLayout.error = "Name is required"
                false
            }
            phone.isEmpty() -> {
                phoneLayout.error = "Phone number is required"
                false
            }
            address.isEmpty() -> {
                addressLayout.error = "Address is required"
                false
            }
            birthdate.isEmpty() -> {
                birthdateLayout.error = "Birthdate is required"
                false
            }
            else -> true
        }
    }
    
    private fun updateProfile() {
        loader.apply {
            setMessage("Updating profile...")
            setCanceledOnTouchOutside(false)
            show()
        }
        
        val userMap = hashMapOf<String, Any>(
            "name" to nameInput.text.toString().trim(),
            "phonenumber" to phoneInput.text.toString().trim(),
            "address" to addressInput.text.toString().trim(),
            "birthdate" to birthdateInput.text.toString().trim()
        )
        
        // Add profile image path if available
        currentPhotoFile?.let {
            userMap["profileImagePath"] = it.absolutePath
        }
        
        userRef.updateChildren(userMap).addOnCompleteListener { task ->
            loader.dismiss()
            if (task.isSuccessful) {
                Toast.makeText(this, "Profile updated successfully", Toast.LENGTH_SHORT).show()
                finish()
            } else {
                Toast.makeText(
                    this,
                    "Error updating profile: ${task.exception?.message}",
                    Toast.LENGTH_SHORT
                ).show()
            }
        }
    }
    
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                finish()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }
}
